### ISSUE TITLE
Help with Humanoid Grasp Task

### ISSUE THREAD

**Issue by Lrogic** [haosulab/ManiSkill#1465]
Hello,

I’m trying to train the Unitree G1 humanoid to grasp and lift an object stably, but I have not been able to get a reliable grasp policy.

The longer-term goal is:

1. Learn a grasp policy.
2. Use IK to transport the grasped object to a target location.
3. Learn a release policy to place/release the object stably.

I figured out the IK, but now I'm trying to figure out the grasp part. I would appreciate any guidance on whether there are obvious issues in my reward design, observations, reset setup, embodiment/controller choice, or training setup.

I tried two main approaches.

---

## Approach 1: Simple grasp reward

In the first version, the reward is intentionally simple:

* reward for reaching the object
* success reward if the object is grasped for enough consecutive steps

The observations include:

* end-effector pose
* gripper width
* goal/object pose information
* whether the target object is_grasped or not

**More details and the relevant code are included at the bottom of this issue.**

Training setup:

```bash
--total-timesteps 300_000_000
--num-steps 32
--update-epochs 8
--control-mode pd_joint_delta_pos
```

For the embodiment, I subclassed `UnitreeG1UpperBody` to create a G1 variant where the left arm is frozen, but the torso is still allowed to rotate. These PPO parameters were based on the ManiSkill PPO baseline script:
https://github.com/mani-skill/ManiSkill/blob/main/examples/baselines/ppo/baselines.sh

Video of the trained policy:

https://github.com/user-attachments/assets/373f306c-ab00-4a44-8ac8-47f462169d2e

---

## Approach 2: Reach + grasp + lift + stable hold reward

I also tried a more shaped reward with terms for:

* reaching the object
* grasping
* lifting
* holding the object stably
* final success

For this version, I also changed the observations to be relative to the robot root instead of world coordinates, so the policy can hopefully generalize better to new object locations.

For this approach, I also used a different G1 embodiment from the Mink repo. Compared to the previous setup, this embodiment includes 3 additional waist DOFs and 3 additional wrist DOFs. I still froze the left arm in this setup using a passive controller, while allowing the torso/right arm/right wrist to move.

**More details and the relevant code are included at the bottom of this issue.**

Training setup (same as Approach 1):

```bash
--total-timesteps 300_000_000
--num-steps 32
--update-epochs 8
--control-mode pd_joint_delta_pos
```

I tested two reset setups:

1. Standard G1 reset pose, with the arm starting by the side.
2. IK-generated pregrasp reset pose, with the hand starting above the object.

Videos:

Pregrasp reset pose:

https://github.com/user-attachments/assets/3e249663-edab-4df1-857f-1a484ed1b71f

Standard reset pose:

https://github.com/user-attachments/assets/fa8d9b2c-06da-46d7-a84c-012ffbe3f649

---

## Main questions

Is there anything obviously wrong or suspicious about this setup?

In particular, I’m wondering about:

* In both approaches, the policy seems to exploit the reaching reward by hovering near the object without actually learning to grasp it. Would it make sense to zero out or saturate the reaching reward once the hand is within some distance threshold, so the policy is encouraged to discover grasping behavior?
* whether the reward design is reasonable for learning a stable grasp?
* whether the observations are missing something important?
* whether using normalize_action in the controller_configs is important for training?
* should I use ppo_fast or sac instead of ppo? what's the differences?
* It seems in the UnitreeG1PlaceAppleInBowl task that the table height is higher, and also the apple is an easier object to grasp than the jenga block. Could that be part of the issue?

Relevant code snippets are below.

---

### Approach 1: Simple reward and observations

```python
    def compute_dense_reward(self, obs: Any, action: torch.Tensor, info: dict):
        info.setdefault("extra_data", {})
        extra = info["extra_data"]

        reaching_reward = self._compute_reaching_reward(info)
        grasp_success_reward = self.success_bonus * info["success"].float()
        reward = reaching_reward + grasp_success_reward

        extra["reaching_reward"] = reaching_reward
        extra["grasp_success_reward"] = grasp_success_reward
        extra["is_grasped"] = info["is_grasped"].float()
        extra["success"] = info["success"].float()
        extra["stable_grasp_counter"] = self.stable_grasp_counter.float()
        return reward

    def _current_achieved_goal(self) -> torch.Tensor:
        return self.manipulated_object.pose.raw_pose

    def _current_desired_goal(self) -> torch.Tensor:
        return self.manipulated_waypoints[self.current_subgoal_idx]

    def _noisy_achieved_goal(self) -> torch.Tensor:
        """Object pose for observations; true pose is used for rewards and success."""
        obj_obs = self._current_achieved_goal()
        if self.obj_noise <= 0:
            return obj_obs
        noise = torch.rand_like(obj_obs) * (self.obj_noise * 2) - self.obj_noise
        return obj_obs + noise


    def _get_obs_extra(self, info: dict):
        gripper_width_proxy = self.agent.right_hand_dist_to_open_grasp().unsqueeze(-1)
        obs = dict(
            ee_pose=self.agent.right_tcp.pose.raw_pose,
            gripper_width=gripper_width_proxy,
        )

        if "state" in self.obs_mode:
            achieved_goal = self._noisy_achieved_goal()
            desired_goal = self._current_desired_goal()
            goal_position_diff = desired_goal[:, :3] - achieved_goal[:, :3]
            goal_rotation_diff = torch.abs(
                self._quaternion_angle(desired_goal[:, 3:7], achieved_goal[:, 3:7])
            ).unsqueeze(-1)
            goal_obs = dict(
                goal_position_diff=goal_position_diff,
                goal_rotation_diff=goal_rotation_diff,
            )
            if self.include_absolute_goal_obs:
                goal_obs["achieved_goal"] = achieved_goal
                goal_obs["desired_goal"] = desired_goal
            obs.update(goal_obs)

        if self.include_grasp_obs and "is_grasped" in info:
            obs["is_grasped"] = info["is_grasped"]

        if self.include_task_phase_obs:
            obs["task_phase"] = torch.nn.functional.one_hot(
                self.task_phase, num_classes=TaskPhase.COUNT
            ).float()

        return obs
```

---

### Approach 2: More shaped reward and root-relative observations

```python
    def _get_obs_extra(self, info: dict):
        root_inv = self.agent.robot.root.pose.inv()
        ee_pose = (root_inv * self.agent.right_tcp.pose).raw_pose
        object_pose = (root_inv * self.manipulated_object.pose).raw_pose
        object_pose_palm = (
            self.agent.right_tcp.pose.inv() * self.manipulated_object.pose
        ).raw_pose
        gripper_width = self.agent.right_hand_dist_to_open_grasp().unsqueeze(-1)

        obs = dict(
            ee_pose=ee_pose,
            gripper_width=gripper_width,
            object_pose=object_pose,
            object_pose_palm=object_pose_palm,
        )
        if self.include_grasp_obs and "is_grasped" in info:
            obs["is_grasped"] = info["is_grasped"].float().unsqueeze(-1)
        return obs

    def compute_dense_reward(self, obs: Any, action: torch.Tensor, info: dict):
        info.setdefault("extra_data", {})
        extra = info["extra_data"]

        is_grasped = info["is_grasped"].float()
        is_lifted = info["is_lifted"].float()
        is_static = info["is_static"].float()

        object_to_palm_dist = info["object_to_palm_dist"]
        reaching_reward = 1.0 - torch.tanh(self.reaching_scale * object_to_palm_dist)

        grasp_reward = is_grasped
        lifted_progress = torch.clamp(
            (info["object_height"] - self.initial_object_z) / self.lift_height_threshold,
            0.0,
            1.0,
        )
        lift_reward = is_grasped * lifted_progress

        linear_static_reward = 1.0 - torch.tanh(
            5.0 * torch.linalg.norm(self.manipulated_object.linear_velocity, dim=1)
        )
        angular_static_reward = 1.0 - torch.tanh(
            2.0 * torch.linalg.norm(self.manipulated_object.angular_velocity, dim=1)
        )
        static_reward = linear_static_reward * angular_static_reward

        stable_hold_reward = is_grasped * is_lifted * static_reward

        xy_penalty = torch.clamp(info["object_xy_displacement"] - self.xy_tolerance, min=0.0)
        orientation_diff = torch.abs(
            self._quaternion_angle(
                self.initial_object_quat,
                self.manipulated_object.pose.q,
            )
        )
        orientation_penalty = torch.clamp(orientation_diff - self.orientation_tolerance, min=0.0)

        success_reward = info["success"].float()

        reward = (
            self.w_reach * (1.0 - is_grasped) * reaching_reward
            + self.w_grasp * grasp_reward
            + self.w_lift * lift_reward
            + self.w_hold * stable_hold_reward
            + self.w_success * success_reward
            - self.w_xy * xy_penalty
            - self.w_ori * orientation_penalty
        )
        return reward

```
